lib/Serialization/DeserializeSIL.cpp

```diff
@@ -595,7 +597,8 @@ SILDeserializer::readSILFunctionChecked(DeclID FID, SILFunction *existingFn,
     for (auto ID : SemanticsIDs) {
       fn->addSemanticsAttr(MF->getIdentifierText(ID));
     }
-
+    if (!hasQualifiedOwnership)
```

Comment:
@atrick this is the important part of the change. The rest of the hunks are just some small cleanups.

Reply:
It looks to me like `setOwnershipEliminated` is called twice now for the same function. It seems like there should be one place where it's done. Or did I misread something?

Reply:
Its actually more confusing than that. The thing that you are talking about is one that is only used to set ownership if we are deserializing a body. The bug here is that if we only want to create a declaration (something that happens often times when linking sil code), we get into the bad state where we have an existing function with the ossa flag set, but the body says it shouldn't have ownership. (I originally found this inconsistency when trying to add an assert in the if block above that those were always consistent).

That being said... I was thinking about this a little bit. If you think about it ownership is really a property of the body of a function, and thus not the declaration. That makes me wonder if we really we want a tri-state here where if one has a declaration, you get none. Then all of these weird issues go away. I am still a little nervous about people caching info/assumptions around what is in ownership/etc... but that is just normal analysis invalidation stuff.

Your thoughts?

Reply:
Also, this code is spaghetti code... the control flow is all jumbled... = /.

Reply:
I'm questioning why `if (!hasQualifiedOwnership) setOwnershipEliminated()` isn't just done unconditionally before returning the declaration.

Reply:
I was keeping the overall form of the code the same. I am happy to rework this a little.

Reply:
I'm not trying to impose extra work. I'm genuinely confused by the code. Seems like there should be one place to handle both decls and defs or it should be made explicit in each case what is being handled. The potential clash between decl and def properties is really hidden by the code.